src/tokenizer/shingle_filter.rs

```diff
@@ -0,0 +1,207 @@
+use super::{BoxTokenStream, Token, TokenFilter, TokenStream};
+
+#[derive(Clone)]
+pub struct FixedShingleFilter {
+    size: usize,
+    separator: char,
+}
+
+impl FixedShingleFilter {
+    pub fn new(size: usize) -> Self {
+        FixedShingleFilter::new_with_separator(size, '_')
+    }
+
+    pub fn new_with_separator(size: usize, separator: char) -> Self {
+        FixedShingleFilter { size, separator }
+    }
+}
+
+impl TokenFilter for FixedShingleFilter {
+    fn transform<'a>(&self, token_stream: BoxTokenStream<'a>) -> BoxTokenStream<'a> {
+        BoxTokenStream::from(ShingleFilterStream {
+            size: self.size,
+            separator: self.separator,
+            buffer: String::with_capacity(100),
+            pendings: Vec::with_capacity(self.size - 1),
+            tail: token_stream,
+            original_token_position: 0,
+        })
+    }
+}
+
+pub struct ShingleFilterStream<'a> {
+    size: usize,
+    separator: char,
+    buffer: String,
+    pendings: Vec<Token>,
+    tail: BoxTokenStream<'a>,
+    original_token_position: usize,
+}
+
+impl<'a> ShingleFilterStream<'a> {
+    fn is_window_filled(&self) -> bool {
+        self.pendings.len() == self.size - 1
+    }
+
+    fn push_token(&mut self, mut token: Token) {
+        if self.is_window_filled() {
+            self.pendings.remove(0);
+        }
+        token.position = self.original_token_position;
+        self.pendings.push(token);
+    }
+}
+
+fn make_shingle(token: &mut Token, pendings: &Vec<Token>, buffer: &mut String, separator: char) {
+    buffer.clear();
+    buffer.push_str(token.text.as_str());
+
+    // set text
+    token.text.clear();
+    for pend_token in pendings {
+        token.text.push_str(pend_token.text.as_str());
+        token.text.push(separator);
+    }
+    token.text.push_str(buffer.as_str());
+
+    // set offset and position
+    if let Some(first_token) = pendings.get(0) {
+        token.offset_from = first_token.offset_from;
+        token.position = first_token.position;
+    }
+}
+
+impl<'a> TokenStream for ShingleFilterStream<'a> {
+    fn advance(&mut self) -> bool {
+        while !self.is_window_filled() {
+            if self.tail.advance() {
+                self.push_token(self.tail.token().clone());
+                self.original_token_position += 1;  // This is not compatible with StopWordFilter
+            } else {
+                return false;
+            }
```

Comment:
nitpick:
The whole codebase is not consistent but we try to use early returns these days..

```suggestion
            if !self.tail.advance() {
                return false;
            }
            self.push_token(self.tail.token().clone());
            self.original_token_position += 1;  // This is not compatible with StopWordFilter
```